[ENH] Minigallery can take arbitrary files/glob patterns as input (#1226)
sphinx-gallery/sphinx-gallery: doc/configuration.rst

For such a mini-gallery, specifying a custom heading message is recommended
because the default message is vague: "Examples of one of multiple objects".

Create mini-galleries using filepaths
"""""""""""""""""""""""""""""""""""""
Sometimes you may want to explicitly create a mini-gallery for files that may be 
in the same location but not have functions in common, for example a set of 
tutorials. The mini-gallery directive supports passing in a string path and 
a glob-style string to a set of files. 

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
   :add-heading: Mini-gallery using paths
   :heading-level: -

The order of the minigallery file based entries can be set using the 
``minigallery_sort_order`` configuration key. This configuration key supports
any callable that is a valid *key* argument to ``sorted``::

    sphinx_gallery_conf = {
    ...,
    'minigallery_sort_order' : lambda x: return len(x)%4,
    ... 

>     }

Reviewer: Instead of a trivial example here, could you make one that actually does something interesting related to the example above with `numpy.exp ../examples/...`? I think it will be easier for people to understand the use case

For such a mini-gallery, specifying a custom heading message is recommended
because the default message is vague: "Examples of one of multiple objects".

Create mini-galleries using filepaths
"""""""""""""""""""""""""""""""""""""
Sometimes you may want to explicitly create a mini-gallery for files that may be 
in the same location but not have functions in common, for example a set of 
tutorials. The mini-gallery directive supports passing in a string path and 
a glob-style string to a set of files. 

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
   :add-heading: Mini-gallery using paths
   :heading-level: -

The order of the gallery entries is determined by the argument order. The ordering of 
files matching a specific glob pattern, e.g. ``../examples/plot_4*`` can be set using the 
``minigallery_sort_order`` configuration key. This configuration key supports
any callable that is a valid *key* argument to ``sorted``::

    sphinx_gallery_conf = {
    ...,
    # put duplicate figure at end 

>     'minigallery_sort_order' : lambda path: not path.startswith('plot_4b')

Reviewer: Wouldn't this be better, i.e., sort first based on "plot_4b"-ness then on the filename itself?
```suggestion
    'minigallery_sort_order' : lambda path: (not path.startswith('plot_4b'), path)
```

For such a mini-gallery, specifying a custom heading message is recommended
because the default message is vague: "Examples of one of multiple objects".

Create mini-galleries using file paths
""""""""""""""""""""""""""""""""""""""
Sometimes you may want to explicitly create a mini-gallery for files that may be 
in the same location but not have functions in common, for example a set of 

> tutorials. The mini-gallery directive supports passing in a string path and 

Reviewer 2: Maybe this could be clearer?

a glob-style string to a set of files. 

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
   :add-heading: Mini-gallery using paths
   :heading-level: -

The ordering of thumbnails can be set using the ``minigallery_sort_order`` 
configuration key. This configuration key supports any callable that is a valid 
*key* argument to ``sorted`` and is provided two parameters to sort on, 

> the input ``obj`` and the filepath ``path``:

Reviewer 2: > Also, should I just keep it simple and only sort on paths?

Sorry I don't follow. As in only provide `path` and not `obj`?

.. code-block:: rst
    
    sphinx_gallery_conf = {
    ...,
     "minigallery_sort_order": lambda x: (

>         not (x[0] == "numpy.exp"),

Reviewer 2: Oh maybe an explanation about what we are doing (i'm probably just bad with `key`) ?

        not x[1].startswith("plot_4b"),

>         x[1],

Reviewer: Okay one last very annoying (and tricky!) thing -- it's important that whatever is in the `sphinx_gallery_conf` have a stable `__repr__` from run to run, otherwise rebuilds will happen for all files on sequential calls to `sphinx_build` even if no files have changed. Unfortunately, we can't guarantee that `lambda`'s `__repr__` will be the same between subsequent runs of Sphinx. So really this needs to be something like:
```
class MinigallerySortOrder:

    def __repr__(self):
         return "MinigallerySortOrder"

    def __call__(self, fname):
        return ...
```
then you pass `"minigallery_sort_order": MinigallerySortOrder(),` as the conf entry. See the `warning` in https://sphinx-gallery.github.io/stable/configuration.html#sorting-gallery-subsections for example

For such a mini-gallery, specifying a custom heading message is recommended
because the default message is vague: "Examples of one of multiple objects".

.. _file_based_minigalleries:

Create mini-galleries using file paths
======================================

Sometimes you may want to explicitly create a :class:`mini-gallery <sphinx_gallery.directives.MiniGallery>` 

> for files that may be in the same location but not have functions in common, for example a set of 

Reviewer 2: nit - line length?

> tutorials. The mini-gallery directive therefore supports passing in:

Reviewer 2: ```suggestion
tutorials. The mini-gallery directive therefore supports passing in
a list (space separated) of any of the following:
```

* full qualified name of object
* pathlike strings to example Python files
* glob-style pathlike strings to example Python files

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
    :add-heading: Mini-gallery using paths
   :heading-level: ^

Sort mini-gallery thumbnails from files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The ordering of thumbnails can be set using the ``minigallery_sort_order`` 

Reviewer 2: Can we make explicit that all minigalleries created with the `minigallery` directive will use the single `minigallery_sort_order` config value?

Reviewer 2: Ohh sorry I was not clear and got mixed up. I did not mean to include backreferences. I mean to make it clear that the same `minigallery_sort_order` is used for all file minigalleries , but now that I think about it I don't think you'd ever want use different sort keys for different galleries. Resolving.

configuration key. Minigallery provides two values to sort on:

* input: object name, file name, or glob string

> * path: backreferences ``{input}.example`` or ``{gallery_example}.py`` file

Reviewer 2: ```suggestion
* path: backreferences ``{input}.example`` or ``{gallery_example}.py`` file string
```

> The sorting key object is passed to :py:func:`sorted` must have the following:

Reviewer 2: ```suggestion
The sorting key object passed to :py:func:`sorted` must have the following:
```

    * ``__repr__`` that is :ref:`stable across runs <sort-warning>`
    * ``__call__`` that takes one argument. 

The sorting method :class:`sphinx_gallery.sorting.FunctionSortKey` can be used 
to wrap arbitrary sorting key functions:

.. code-block:: python
    
    sphinx_gallery_conf = {
    #...,
    "minigallery_sort_order": FunctionSortKey(lambda x: (
        not (x[0] == "numpy.exp"), # input
        not x[1].startswith("plot_4b"), # path
        x[1],
    )),
    #... 
    }

> Sorting is not supported for thumbnails generated from references, as described in 
> :ref:`references_to_examples`.

Reviewer 2: ```suggestion
Sorting is not supported for a set/subset of thumbnails generated from
 :ref:`backreferences <references_to_examples>` i.e., the thumbnails
 linked to a qualified object name input.
```

I get what this means now. It's hard to document this clearly, I tried my best.

instead of any character, if the name is unambiguous you can also write
``pyplot.show`` or just ``show``).

.. _file_based_minigalleries:

Create mini-galleries using file paths
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Sometimes you may want to explicitly create a 
:class:`mini-gallery <sphinx_gallery.directives.MiniGallery>` using files that 
do not have functions in common, for example a set of tutorials. The 
mini-gallery directive therefore also supports passing in:

* pathlike strings to example Python files
* glob-style pathlike strings to example Python files

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
    :add-heading: Mini-gallery using paths
    :heading-level: ^

Sort mini-gallery thumbnails from files
"""""""""""""""""""""""""""""""""""""""

> The mini-gallery directive generates a gallery thumbnails corresponding to the 

Reviewer 2: "a gallery of thumbnails" ? or "generates gallery thumbnails"

instead of any character, if the name is unambiguous you can also write
``pyplot.show`` or just ``show``).

.. _file_based_minigalleries:

Create mini-galleries using file paths
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Sometimes you may want to explicitly create a 
:class:`mini-gallery <sphinx_gallery.directives.MiniGallery>` using files that 
do not have functions in common, for example a set of tutorials. The 
mini-gallery directive therefore also supports passing in:

* pathlike strings to example Python files
* glob-style pathlike strings to example Python files

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
    :add-heading: Mini-gallery using paths
    :heading-level: ^

> 

Reviewer 2: Ah sorry, I didn't see it in context. Ignore in that case, sorry!

``pyplot.show`` or just ``show``).

.. _file_based_minigalleries:

Create mini-galleries using file paths
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Sometimes you may want to explicitly create a 
:class:`mini-gallery <sphinx_gallery.directives.MiniGallery>` using files that 
do not have functions in common, for example a set of tutorials. The 
mini-gallery directive therefore also supports passing in:

* pathlike strings to sphinx gallery example files
* glob-style pathlike strings to sphinx-gallery example files

.. code-block:: rst

    .. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*

.. minigallery:: numpy.exp ../examples/plot_0_sin.py ../examples/plot_4*
    :add-heading: Mini-gallery using paths
    :heading-level: ^

Sort mini-gallery thumbnails from files
"""""""""""""""""""""""""""""""""""""""

The mini-gallery directive generates a gallery of thumbnails corresponding to 
the input file strings or object names. The ordering of these thumbnails can be 
set using the ``minigallery_sort_order`` configuration key. The minigallery 
directive passes the following information to :py:func:`sorted`:

> * sphinx-gallery :ref:`sorting key <sort-warning>` method

Reviewer 2: I don't quite follow this. I thought the `sorted` function got the `minigallery_sort_order` function?
